docs/ds20.md

```diff
@@ -0,0 +1,111 @@
+
+    [USB\CLASS_FF&SUBCLASS_42&PROT_03]
+    Plugin = fastboot
+
+#### Microsoft OS Descriptors 1.0
+
+The [Microsoft OS Descriptors 1.0 Specification](https://docs.microsoft.com/en-us/windows-hardware/drivers/usbcon/microsoft-os-1-0-descriptors-specification) defines a device-specific variable-length metadata block of `CompatibleID`:`SubCompatibleID` on a string index of `0xEE` where the `CompatibleID` and `SubCompatibleID` are also both hardcoded at 8 bytes.
+
+We could squeeze `FWUPDPLU` or `FWUPDFLA` as the `CompatibleID`, but we couldn’t squeeze the plugin name (e.g. `logitech-bulkcontroller`) or the GUID (16 bytes) in an 8 byte `SubCompatibleID`.
+Some non-compliant devices also don’t respond well (as in, they hang, and stop working...) when probing this specific string index.
+
+#### Microsoft OS Descriptors 2.0
+
+The [Microsoft OS Descriptors 2.0 Specification](https://docs.microsoft.com/en-us/windows-hardware/drivers/usbcon/microsoft-os-2-0-descriptors-specification) is more helpful as it defines a new device capability that can return a variable length chunk of data, using a UUID as a key.
+We do need to use the `BOS` *“Binary Object Store”* descriptor, which is only supported for devices using USB 2.1 and newer, although that’s probably the vast majority of devices now.
+Usually the BOS descriptor is used for Wireless USB details, USB 2.0 extensions, SuperSpeed USB connection details and a Container ID.
+
+Usefully, the BOS descriptor does give us is the ability to use the platform capability descriptor, which is `bDevCapabilityType=0x05`. For UUID `D8DD60DF-4589-4CC7-9CD2-659D9E648A9F`, this is identified as a structured blob of data Windows usually uses to put workarounds like the suspend mode of the device and that kind of thing.
+
+We could certainly create a new `bDevCapabilityType` which would allow us to store a binary blob (e.g. `Plugin=foobarbaz\nFlags=QuirkValueHere\n`) but that would be out-of-specification and difficult for vendors to implement.
+
+## Proposal
+
+Devices create a new platform device capability BOS descriptor with `bDevCapabilityType=0x05` and a fwupd-specific UUID, specifically `010aec63-f574-52cd-9dda-2852550d94f0`.
+Using this custom UUID will ensure that Microsoft Windows does not try to parse the capability descriptor as a *Microsoft OS 2.0 descriptor set*.
+
+## Implementation
+
+Create a BOS DS20 descriptor such as:
+
+    1C 10 05 00 63 ec 0a 01 74 f5 cd 52 9d da 28 52 55 0d 94 f0 05 08 01 00 20 00 2a 00
+    |  |  |  |  \---------------------------------------------/ \--------/  \---/ |  |
+    |  |  |  |                ^-PlatformCapabilityUUID            |   wLength-^   |  |
+    |  |  |  ^-bReserved                                 dwVersion^   bVendorCode-^  |
```

Comment:
I noticed you didn't explain `bVendorCode`.  What about using `bVendorCode` to designate a numerical value corresponding to a specific USB plugin or group of plugins?  We are of course limited to 255 USB plugins, but that _seems_ reasonable to me.

Reply:
It's there:

    bRequest: {value of bVendorCode in the BOS descriptor}

It's explained in much more detail in https://docs.microsoft.com/en-us/windows-hardware/drivers/usbcon/microsoft-os-2-0-descriptors-specification which is probably some good bedtime reading. It's way more flexible as it defines the bRequest so you can return almost unlimited data if you want to.

Reply:
Yeah I saw that, but that's how the "client" responds, not the value that the vendor should be placing _into_ the descriptor.  So I was thinking we could prescribe the values there.

The Microsoft one is tangential, remember it's a fwupd spec!

Reply:
Right, it's tangential -- but I think it makes a lot of sense to copy the API design -- we need to make this as least weird as possible for vendors. They're already familiar with the _control transfer with buffer of data_ thing and it's useful to have similar version semantics.

Reply:
In that case we need to specify what exactly we **want** them to put in that field in this spec, even if it's similar to the other.

Reply:
Back a bit; the descriptor just says *hit this control transfer to get the raw data` -- it's literally just the UTF-8 data rather than any specific encoding or framing.

Reply:
But who decides the value should be `0x2a`?  I'm Mario's new dongle shop and I make a new device.  How do I know what to put in that field?
That's what is missing from the spec.

Reply:
Gotcha. What about `A suitable bVendorCode should be chosen that is not used for existing device operation`?

Reply:
Yeah that's fine

Reply:
Fixed.